File: python/src/middleware/audit.py
"""Audit logging middleware — structured JSON logs for every request.

Records user_id, action (HTTP method), resource (path), timestamp, and request_id
for compliance and incident response. Logs are emitted as structured JSON via structlog.
"""

import time
from datetime import datetime, timezone

import structlog
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.requests import Request

logger = structlog.get_logger("audit")


class AuditMiddleware(BaseHTTPMiddleware):
    """Logs an audit trail entry for every request with user context when available."""

    async def dispatch(self, request: Request, call_next):
        start_time = time.monotonic()

        response = await call_next(request)

        duration_ms = round((time.monotonic() - start_time) * 1000, 2)

        # Extract user_id from request state (set by auth middleware).
        # Falls back to "anonymous" for unauthenticated endpoints like /healthz.
        user_id = "anonymous"
        if hasattr(request.state, "user") and request.state.user:
            user_id = request.state.user.get("sub", "anonymous")

        # Extract request_id from request state (set by RequestIDMiddleware).
        request_id = getattr(request.state, "request_id", "unknown")

        # Client IP for audit trail — use forwarded header if behind a proxy.
        client_ip = request.headers.get("X-Forwarded-For", "").split(",")[0].strip()
        if not client_ip:
            client_ip = request.client.host if request.client else "unknown"

        logger.info(
            "audit",
            user_id=user_id,
            action=request.method,
            resource=request.url.path,
            query=str(request.query_params) if request.query_params else None,
            status_code=response.status_code,
            duration_ms=duration_ms,
            client_ip=client_ip,
            request_id=request_id,
            timestamp=datetime.now(timezone.utc).isoformat(),
            user_agent=request.headers.get("User-Agent", ""),
        )

        return response

File: go/middleware/audit.go
package middleware

import (
	"log/slog"
	"net/http"
	"time"
)

// AuditLog returns middleware that logs a structured audit trail entry for every
// request. Each entry includes user_id, action (method), resource (path),
// status code, duration, client IP, request ID, and timestamp.
//
// When auth middleware has run, the user is extracted from context. Otherwise
// the entry logs "anonymous" — this is expected for unauthenticated endpoints
// like /healthz.
func AuditLog(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		start := time.Now()

		// Wrap ResponseWriter to capture the status code.
		sw := &statusWriter{ResponseWriter: w, status: http.StatusOK}

		next.ServeHTTP(sw, r)

		duration := time.Since(start)

		// Extract user from context (set by RequireAuth middleware).
		userID := "anonymous"
		if user := UserFromContext(r.Context()); user != nil {
			userID = user.Subject
		}

		// Extract request ID from context (set by RequestID middleware).
		requestID := RequestIDFromContext(r.Context())

		// Client IP for audit trail.
		clientIP := extractIP(r)

		slog.Info("audit",
			"user_id", userID,
			"action", r.Method,
			"resource", r.URL.Path,
			"query", r.URL.RawQuery,
			"status_code", sw.status,
			"duration_ms", duration.Milliseconds(),
			"client_ip", clientIP,
			"request_id", requestID,
			"timestamp", start.UTC().Format(time.RFC3339),
			"user_agent", r.UserAgent(),
		)
	})
}

// statusWriter wraps http.ResponseWriter to capture the response status code.
type statusWriter struct {
	http.ResponseWriter
	status      int
	wroteHeader bool
}

func (w *statusWriter) WriteHeader(code int) {
	if !w.wroteHeader {
		w.status = code
		w.wroteHeader = true
	}
	w.ResponseWriter.WriteHeader(code)
}

func (w *statusWriter) Write(b []byte) (int, error) {
	if !w.wroteHeader {
		w.wroteHeader = true
	}
	return w.ResponseWriter.Write(b)
}
